source/s3_meta_request.c

```diff
@@ -168,19 +242,20 @@ static void s_s3_request_on_finish(
             Py_XDECREF(header_list);
             goto done;
         }
-        struct aws_byte_buf *error_body = meta_request_result->error_response_body;
-        result = PyObject_CallMethod(
-            request_binding->self_py,
-            "_on_finish",
-            "(iOy#)",
-            meta_request_result->error_code,
-            header_list,
-            (const char *)(error_body->buffer),
-            (Py_ssize_t)error_body->len);
+    }
+    if (meta_request_result->error_response_body) {
+        error_body = *(meta_request_result->error_response_body);
+    }
+    result = PyObject_CallMethod(
+        request_binding->self_py,
+        "_on_finish",
+        "(iOy#)",
+        meta_request_result->error_code,
+        header_list,
+        (const char *)(error_body.buffer),
+        (Py_ssize_t)error_body.len);
+    if (header_list != Py_None) {
         Py_XDECREF(header_list);
```

Comment:
to simplify the logic around whether or not you need to decref header_list, I'd suggest:
```
PyObject *result = NULL;
PyObject *header_list = NULL; // move this up to top of gil acquire block
...
PyObject_CallMethod(
    ...,
    header_list ? header_list : Py_NONE,
    ...);
...

done:
Py_XDECREF(header_list); // just one XDECREF(header_list)at end of gil block
```